### PR TITLE
#1240 Inbound shipment service charge tax pen icon glitch with calculating Service price total on fly

### LINKED ISSUE #1240: Inbound invoice: Updating service tax amount > Error
## What went wrong? 😲

Updating service tax amount in Inbound shipment > Gives error.

Inbound error

## Expected behaviour 🤔

Must be handled

## How to Reproduce 🔨

Steps to reproduce the behaviour:

1. Any Inbound shipment, Any status not `Confirm Verified`
2. Click on 'More'
3. Service charges section
4. Click on Tax 'pen' icon
5. Set the tax amount
6. Click OK
7. See error

## Your environment 🌱
 - Version [1.2.3, current develop]
 - Platform [browser, windows desktop, server]

### PR DESCRIPTION
**Pull request by bijaySussol**
Fixes #1240 

# 👩🏻‍💻 What does this PR do? 
This will disable the Tax pen icon disable for the Service charges section

Screen Shot 2023-03-09 at 2 03 10 PM

## 💌 Any notes for the reviewer?

~Considering the workflow of Outbound > Service charges section, This does not get enabled at all so considering that, thinking is, The service charges are yet to get applied.~
~So, Have disabled the Tax pen icon on Inbound shipment as well.~

~Also, It seems, Service charge is per line, Does not make sense though, and there is no handler in Backend to the best of my understanding.~

The issue is the GraphQl Typename.

### CODE REVIEW

**Comment by anildahalsussol** [comment]
Even if we add a service charge then we are not allowed to edit tax using that editing pen. But we can do it from the service charge. Not sure do we need that if its not in use?
Screenshot 2023-03-10 at 10 47 52

**Comment by bijaySussol** [comment]
> Even if we add a service charge then we are not allowed to edit tax using that editing pen. But we can do it from the service charge. But in outbound, we are allowed to edit... Screenshot 2023-03-10 at 10 47 52

Fixed by [ca6e220](https://github.com/openmsupply/open-msupply/pull/1311/commits/ca6e220ea96fd3efd4a23a1195db48b1b1814718)
Thanks

File: client/packages/invoices/src/InboundShipment/DetailView/SidePanel/PricingSection.tsx [hunk 1]
```diff
@@ -98,7 +98,7 @@ export const PricingSectionComponent = () => {
           <PanelLabel>{`${t('heading.tax')} ${Formatter.tax(tax)}`}</PanelLabel>
           <PanelField>
             <TaxEdit
-              disabled={false}
+              disabled={!stockLines?.length}
```
Comment (mark-prins) [inline client/packages/invoices/src/InboundShipment/DetailView/SidePanel/PricingSection.tsx:101]: I think that if the shipment is verified, we shouldn't be able to edit service changes or tax details. This is how mSupply works and I'm taking the logic partly from there 🤷

**Review by mark-prins — COMMENTED** [review]
changes are good - and it's working for me!
I have a wider question - I think that we shouldn't be able to edit verified shipments; including tax and charges.

Extra question - I had to add the service charge to my master list in order to make it available. and I think I shouldn't have to do that. Would require updating the item filter.. thoughts?

**Comment by andreievg** [comment]
> Extra question - I had to add the service charge to my master list in order to make it available. and I think I shouldn't have to do that. Would require updating the item filter.. thoughts?

I think we needed to make a KDD for this when the decision was made, from memory there were some solid reasons: https://github.com/openmsupply/open-msupply/issues/1323

**Comment by anildahalsussol** [comment]
Few things:
1. While adding two service items tax sum up can go more than 100% because we can put value in tax more than 100% but while editing tax manually it doesn't allow us to put more than 100%.
Screenshot 2023-03-13 at 09 28 06

2. For stock charges if we don't have cost then should we allow to add tax? Coz total seems to be wrong while checking cost+tax = 0.
Screenshot 2023-03-13 at 09 22 51

**Comment by anildahalsussol** [comment]
And initially total while adding service charges is also not working. Do we fix in the same issue or we can make a new issue if needed?
Screenshot 2023-03-13 at 09 38 16

**Comment by bijaySussol** [comment]
> I have a wider question - I think that we shouldn't be able to edit verified shipments; including tax and charges.

Thanks.
Fixed by: [8f50ebd](https://github.com/openmsupply/open-msupply/pull/1311/commits/8f50ebdb752388e9f457833fecadca2c22673d37)

File: client/packages/invoices/src/InboundShipment/DetailView/SidePanel/PricingSection.tsx (hunk 1 above)
Comment (bijaySussol) [reply client/packages/invoices/src/InboundShipment/DetailView/SidePanel/PricingSection.tsx]: Agreed. Fixed by: [8f50ebd](https://github.com/openmsupply/open-msupply/pull/1311/commits/8f50ebdb752388e9f457833fecadca2c22673d37)

**Comment by bijaySussol** [comment]
> Extra question - I had to add the service charge to my master list in order to make it available. and I think I shouldn't have to do that. Would require updating the item filter.. thoughts?

Since in desktop, Service item does not have a joins and even if it have (which i think it does not), It will shared across all the existing stores. So, For service items, i think, there should not an ISJ query.

Checking: How is service item are treated on omsupply? Treated as `[transact]service_price`  or a lines?

ATM, to the best of the understanding, It is now treated as a lines which then brings me a question, And adding any new service items from Central server > Attaching to master list (proper one) > It does make it to the remote server and can be seen, But it does not allowed to get issued as a line not also as a service lines. It is always the default one i.e `Service item` which is set in a Preference window on central server, so this is how it is?

**Comment by bijaySussol** [comment]
> 1. While adding two service items tax sum up can go more than 100% because we can put value in tax more than 100% but while editing tax manually it doesn't allow us to put more than 100%.

I agree, having 2 different ways sometime could be mis-leading.
From Tax model dialogue > It is specifically mentioned to 100, but not on service charge model.
Considering the central server > Tax % can set any percentage, so not sure what is the idea.

Screen Shot 2023-03-13 at 1 42 57 PM

**Comment by bijaySussol** [comment]
> 3\. For stock charges if we don't have cost then should we allow to add tax? Coz total seems to be wrong while checking cost+tax = 0.

There had been no restriction on central server so far so probably should be OK, not sure the idea on remote server.
Screen Shot 2023-03-13 at 1 45 06 PM

Probably related to this comment: https://github.com/openmsupply/open-msupply/pull/1311#issuecomment-1465648278

**Comment by bijaySussol** [comment]
> And initially total while adding service charges is also not working. Do we fix in the same issue or we can make a new issue if needed?

Agreed.
I do not mind fixing with this issue, don't mind with new issue.

**Comment by mark-prins** [comment]
> > And initially total while adding service charges is also not working. Do we fix in the same issue or we can make a new issue if needed?
> 
> Agreed. I do not mind fixing with this issue, don't mind with new issue as well

I think fix in this issue.  I agree with Anil's other two as well.. would you mind doing all of these? If you're busy though, just make a new issue and we can merge this one in.

- Calculate total when service charges are added
- disable the stock charges tax entry when the total stock charges are `$0`
- limit the tax percentage to `100` in the service charges modal

**Comment by bijaySussol** [comment]
> > > And initially total while adding service charges is also not working. Do we fix in the same issue or we can make a new issue if needed?
> > 
> > 
> > Agreed. I do not mind fixing with this issue, don't mind with new issue as well
> 
> I think fix in this issue. I agree with Anil's other two as well.. would you mind doing all of these? If you're busy though, just make a new issue and we can merge this one in.
> 
> * Calculate total when service charges are added
> * disable the stock charges tax entry when the total stock charges are `$0`

Added 2 issues for those

> * limit the tax percentage to `100` in the service charges modal

Will fix this here
 
> Note: I also made #1351 as I thought that would be a nice improvement too

Thanks will take a look

**Review by mark-prins — APPROVED** [review]
👍  looking good @bijaySussol!

I've added a suggestion in e3a43408cc9973da5f0ce1a976e86515b33f4438 which will add in the request to limit to 100 in the edit modals too. 

when I was doing that, I noticed this:

Screenshot 2023-03-20 at 10 57 27 AM

the tax percentage shows with 2 decimal places in the side panel, but you can enter loads of decimal places when editing - and if you do, the modal shows the last characters, which makes it look like the tax is more than 100!

I've trimmed the tax percentage to 2 decimal places to make things consistent. Hope it's ok to commit to the branch!

**Pull request merged**